tests/kernel/interrupt/src/static_isr.c

```diff
@@ -0,0 +1,196 @@
+	__asm__ volatile("mov %%esp,%0" : "=m"(*reg) : );
+
+#elif defined(CONFIG_ARCH_POSIX)
+	__asm__ volatile("mov %%esp,%0" : "=m"(*reg) : );
+
+#elif defined(CONFIG_RISCV)
+	__asm__ volatile("mv %0, x2\n\t" : "=r"(*reg));
+
+#elif defined(CONFIG_XTENSA)
+	__asm__ volatile("mov %0, sp\n\t" : "=r"(*reg));
+#else
+#endif
+}
+
+#if !defined(CONFIG_XTENSA)
+static uintptr_t handler_test_result;
+static uintptr_t int_0_stack_pointer;
+static uintptr_t int_1_stack_pointer;
+
+
+/* self-defined handler for testing */
+static void test_handler(const void *param)
+{
+	handler_test_result = (uintptr_t)param;
+
+	/* for different ISR, get esp of current thread */
+	if (handler_test_result == ISR0_ARG) {
+		get_stack_addr(&int_0_stack_pointer);
+	}
+
+	if (handler_test_result == ISR1_ARG) {
+		get_stack_addr(&int_1_stack_pointer);
+	}
+}
+
+/**
+ * @brief Test interrupt features
+ *
+ * @details This test case validates four features of interrupt
+ * - The kernel support configuration of interrupts statically at build time.
+ * - The user shall be able to supply a word-sized parameter when configuring
+ *   interrupts, which are passed to the interrupt service routine when the
+ *   interrupt happens.
+ * - The kernel support multiple ISRs utilizing the same function to process
+ *   interrupts, allowing a single function to service a device that generates
+ *   multiple types of interrupts or to service multiple devices.
+ * - The kernel provide a dedicated interrupt stack for processing hardware
+ *   interrupts.
+ *
+ * @ingroup kernel_interrupt_tests
+ */
+void test_isr_static(void)
+{
+	uintptr_t curr_stack_ptr;
+
+	/**TESTPOINT: configuration of interrupts statically at build time */
+	IRQ_CONNECT(TEST_IRQ_0_LINE, TEST_IRQ_PRI, test_handler,
+		       (void *)ISR0_ARG, 0);
+	IRQ_CONNECT(TEST_IRQ_1_LINE, TEST_IRQ_PRI, test_handler,
+		       (void *)ISR1_ARG, 0);
+
+	zassert_not_equal(handler_test_result, ISR0_ARG,
+		"shall not get parameter before interrupt");
+	zassert_not_equal(handler_test_result, ISR1_ARG,
+		"shall not get parameter before interrupt");
+
+	irq_enable(TEST_IRQ_0_LINE);
+	irq_enable(TEST_IRQ_1_LINE);
+
+	trigger_irq(TEST_IRQ_0_LINE);
+	k_busy_wait(MS_TO_US(DURATION));
+
+	/**TESTPOINT: pass word-sized parameter to interrupt */
+	zassert_equal(handler_test_result, ISR0_ARG,
+		"parameter in interrupt is not correct");
+
+	trigger_irq(TEST_IRQ_1_LINE);
+	k_busy_wait(MS_TO_US(DURATION));
+
+	/**
+	 * TESTPOINT: multiple ISRs utilizing the same function to process
+	 * interrupts.
+	 */
+	zassert_equal(handler_test_result, ISR1_ARG,
+		"parameter(%lx) in interrupt is not correct",
+		handler_test_result);
+
+	zassert_true((int_0_stack_pointer - int_1_stack_pointer)
+			< CONFIG_ISR_STACK_SIZE,
+			"different interrupt should use the same stack");
+
+	get_stack_addr(&curr_stack_ptr);
+
+	/**TESTPOINT: dedicated interrupt stack for processing interrupt */
+	zassert_true((int_0_stack_pointer - curr_stack_ptr)
+			> CONFIG_MAIN_STACK_SIZE,
```

Comment:
This doesn't look correct to me.  This is checking that "an address on the interrupt stack" is more than "main stack size" after "an address on this thread's stack".

What if the linker places the interrupt stack **lower** in memory than this stack?  What if the location of curr_stack_ptr is at the high address of the stack (i.e. it's a shallow call depth) and the int_0_stack_pointer is near the bottom, and the interrupt stack is immediately after?

The test here should be that the int_0_stack_pointer lies outside the main stack.  To do that you need to whitebox the thread struct and pull out the stack size and limit fields to check.

Reply:
Got it, just let me think about it~

Reply:
@andyross I make some change of the checking logic, try to check if these two stack are overlap or not. It seems work well except qemu_cortex_a53 and native_posix board. And the stack information of qemu_cortex_a53 are :
 
start 0x400110c, size 0x1000, delta 0x0 (got it from &z_main_thread.stack_info) 

------ 0x400110c0 ----- main thread stack start -----
------ 0x4001107c ------ &dummy in interrupt context
------ 0x40010f4c ------ &dummy in thread context

so it's weird, seems like un-reasonable for me,  sorry, I was stuck here and still working on it.
I appreciate your comment if any, you always help me a lot~